### PR TITLE
pfs-116 FE only for downloading a report

### PR DESCRIPTION
**Pull request by Nicholassully**

### CODE REVIEW

File: adrs/0002-new-finance-back-end.md [hunk 1]
```diff
@@ -15,7 +15,7 @@ the source of truth for payments moving to us from SOP. We are therefore replaci
 ## Decision
 
 This will be a new Front End service following our established Golang FE pattern. The web pages it serves will have two 
-separate functions: Uploading files for processing, and downloading reports. Reporting will be broken out into a separate
+separate functions: Uploading files for processing, and downloading reports. Reportsing will be broken out into a separate
```
Comment (denialanderror) [inline adrs/0002-new-finance-back-end.md:18]: Guessing a search and replace error?

File: internal/api/submit_download.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+		DateField:          dateTransformed,
+		ToDateField:        toDateTransformed,
+		FromDateField:      fromDateTransformed,
+		Email:              emailField,
+	})
+	if err != nil {
+		return err
+	}
+
+	req, err := c.newSiriusRequest(ctx, http.MethodPost, "/downloads", &body)
```
Comment (denialanderror) [inline internal/api/submit_download.go:46]: Nothing is going to be posted to Sirius. The only thing Sirius will be used for is auth.

File: internal/server/server.go [hunk 3]
```diff
@@ -35,6 +40,8 @@ func New(logger *slog.Logger, client ApiClient, templates map[string]*template.T
 	handleMux("GET /uploads", &GetUploadsHandler{&route{client: client, tmpl: templates["uploads.gotmpl"], partial: "uploads"}})
 	handleMux("GET /annual-invoicing-letters", &GetAnnualInvoicingLettersHandler{&route{client: client, tmpl: templates["annual_invoicing_letters.gotmpl"], partial: "annual-invoicing-letters"}})
 
+	handleMux("POST /downloads", &SubmitDownloadHandler{&route{client: client, tmpl: templates["downloads.gotmpl"], partial: "error-summary"}})
```
Comment (denialanderror) [inline internal/server/server.go:43]: Why is this a POST? It is fetching data, not creating it.

File: internal/server/submit_download.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+import (
+	"errors"
+	"github.com/opg-sirius-finance-admin/internal/api"
+	"github.com/opg-sirius-finance-admin/internal/model"
+	"github.com/opg-sirius-finance-admin/internal/util/util"
+	"net/http"
+)
+
+type SubmitDownloadHandler struct {
```
Comment (denialanderror) [inline internal/server/submit_download.go:11]: Strange name. The user is requesting a report to be downloaded as a file, not submitting a download.

**Review by denialanderror — CHANGES_REQUESTED** [review]
This form is meant to select a report type with parameters, which will then generate the report and return it as a file (presuming CSV), but this doesn't appear to do that. We don't have the API for that yet, but that should be what the mock does, so when we do have a solution, we can just plug it in.

File: internal/server/submit_download.go (hunk 4 above)
Comment (Nicholassully) [reply internal/server/submit_download.go]: changed

File: internal/server/server.go (hunk 3 above)
Comment (Nicholassully) [reply internal/server/server.go]: changed

File: internal/api/submit_download.go (hunk 2 above)
Comment (Nicholassully) [reply internal/api/submit_download.go]: Changed

File: internal/api/download.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
+package api
+
+import (
+	"bytes"
+	"encoding/json"
+	"github.com/opg-sirius-finance-admin/internal/model"
+	"net/http"
+)
+
+func (c *ApiClient) Download(ctx Context, reportType string, reportJournalType string, reportScheduleType string, reportAccountType string, reportDebtType string, dateField string, dateFromField string, dateToField string, emailField string) error {
```
Comment (denialanderror) [inline internal/api/download.go:10]: `Field` is redundant naming here. Also, `dateField` needs to be more descriptive to distinguish it from the other date fields.

**Review by denialanderror — CHANGES_REQUESTED** [review]

File: internal/api/download.go [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+	if dateTo != "" {
+		startDateFormatted := model.NewDate(dateTo)
+		toDateTransformed = &startDateFormatted
+	}
+
+	if dateFrom != "" {
+		endDateFormatted := model.NewDate(dateFrom)
+		fromDateTransformed = &endDateFormatted
+	}
```
Comment (hawx) [inline internal/api/download.go:29]: I think you could extract this logic to a `NewDownload` function, which would simplify the `(*ApiClient).Download` method signature a lot

File: internal/api/download.go [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+	req.Header.Set("Content-Type", "application/json")
+
+	resp, err := c.http.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode == http.StatusCreated {
```
Comment (hawx) [inline internal/api/download.go:59]: using switch might read better:

```go
switch resp.StatusCode {
  case http.StatusCreated:
    return nil
  case http.StatusUnauthorized:
    return ErrUnauthorized
  case http.StatusUnprocessableEntity:
    // etc...
}

File: internal/api/download.go [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+		}
+	}
+
+	if resp.StatusCode == http.StatusBadRequest {
+		var badRequests model.BadRequests
+		if err := json.NewDecoder(resp.Body).Decode(&badRequests); err != nil {
+			return err
+		}
+
+		validationErrors := make(model.ValidationErrors)
```
Comment (hawx) [inline internal/api/download.go:80]: (this may be a personal opinion) but I prefer to do `model.ValidationErrors{}` and reserve `make` for only when you want a certain size/capacity

File: internal/api/api_client.go [hunk 9]
```diff
@@ -21,18 +22,22 @@ func (e ClientError) Error() string {
...
+		siriusUrl:  siriusUrl,
+		backendUrl: backendUrl,
+	}, nil
 }
 
 type HTTPClient interface {
 	Do(req *http.Request) (*http.Response, error)
 }
 
 type ApiClient struct {
```
Comment (hawx) [inline internal/api/api_client.go:37]: this is `api.ApiClient`, it would read better named `api.Client` (and then `api.NewClient` above)

File: internal/api/api_client_test.go [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+import (
+	"context"
+	"net/http"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+type MockClient struct {
+	DoFunc func(req *http.Request) (*http.Response, error)
```
Comment (hawx) [inline internal/api/api_client_test.go:12]: this is unused, I'd prefer this over the global `GetDoFunc` below

Comment (hawx) [reply internal/api/api_client_test.go:12]: I'd also recommend looking at mockery for generating these. You can take a look at https://github.com/ministryofjustice/opg-data-lpa-store/pull/253 for how that can be done

File: internal/api/download_test.go [hunk 11]
```diff
@@ -0,0 +1,119 @@
+package api
+
+import (
+	"bytes"
+	"encoding/json"
+	"github.com/opg-sirius-finance-admin/internal/model"
```
Comment (hawx) [inline internal/api/download_test.go:6]: If you haven't already, get https://pkg.go.dev/golang.org/x/tools/cmd/goimports for formatting. Otherwise maybe it needs updating?

File: internal/api/download_test.go [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+			Body:       r,
+		}, nil
+	}
+
+	err := client.Download(getContext(nil), "AccountsReceivable", "", "", "BadDebtWriteOffReport", "", "11/05/2024", "01/04/2024", "31/03/2025", "SomeSortOfEmail@example.com")
+	assert.Equal(t, nil, err)
+}
+
+func TestSubmitDownloadUnauthorised(t *testing.T) {
+	svr := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```
Comment (hawx) [inline internal/api/download_test.go:46]: I'd stick to one style of testing, either use the mock to switch behaviour _or_ do a test server

File: internal/model/date.go [hunk 13]
```diff
@@ -0,0 +1,82 @@
+package model
+
+import (
+	"strings"
+	"time"
+)
+
+type Date struct {
+	Time time.Time
```
Comment (hawx) [inline internal/model/date.go:9]: I think this should be private, so rename to `time` or `t`

File: internal/model/date.go [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+func NewDate(d string) Date {
+	t, err := stringToTime(d)
+	if err != nil {
+		panic(err)
```
Comment (hawx) [inline internal/model/date.go:15]: this should return the error, you want to avoid using `panic` unless you are really sure this is never going to get bad data

File: internal/model/date_test.go [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+			date: NewDate("01/01/0001"),
+			want: true,
+		},
+		{
+			date: Date{},
+			want: true,
+		},
+	}
+	for i, test := range tests {
+		t.Run("Scenario "+strconv.Itoa(i+1), func(t *testing.T) {
```
Comment (hawx) [inline internal/model/date_test.go:98]: I like to make the test cases a `map[string]struct`, then you can use the key to name the scenario

File: internal/model/download.go [hunk 16]
```diff
@@ -0,0 +1,13 @@
+package model
+
+type Download struct {
+	ReportType         string `json:"reportType"`
+	ReportJournalType  string `json:"reportJournalType"`
+	ReportScheduleType string `json:"reportScheduleType"`
+	ReportAccountType  string `json:"reportAccountType"`
+	ReportDebtType     string `json:"reportDebtType"`
+	DateField          *Date  `json:"dateOfTransaction,omitempty"`
```
Comment (hawx) [inline internal/model/download.go:9]: maybe rename to match the json key, this is the only different one

File: internal/model/reports_type.go [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+	case ReportsTypeAccountsReceivable:
+		return "Accounts Receivable"
+	case ReportsTypeDebt:
+		return "Debt"
+	default:
+		return ""
+	}
+}
+
+func (i ReportsType) Key() string {
```
Comment (hawx) [inline internal/model/reports_type.go:39]: does the `String()` method of these things ever vary? I wonder if `Key()` is actually needed, or if it could be `String()` instead

File: internal/server/app_vars.go [hunk 18]
```diff
@@ -10,6 +11,7 @@ type AppVars struct {
 	XSRFToken       string
 	Tabs            []Tab
 	EnvironmentVars EnvironmentVars
+	Errors          model.ValidationErrors
```
Comment (hawx) [inline internal/server/app_vars.go:14]: kind of weird having both `Error` and `Errors` fields, would be nice if they could be combined somehow, maybe? Or rename this to `ValidationErrors` to make it clear it is a different type of thing

File: internal/server/download.go [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+
+import (
+	"errors"
+	"github.com/opg-sirius-finance-admin/internal/api"
+	"github.com/opg-sirius-finance-admin/internal/model"
+	"github.com/opg-sirius-finance-admin/internal/util/util"
+	"net/http"
+)
+
+type DownloadHandler struct {
```
Comment (hawx) [inline internal/server/download.go:11]: I see the other ones are named `{Method}{Action}Handler` so this should be renamed to `GetDownloadHandler` to match

File: internal/server/download_test.go [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+
+	appVars := AppVars{
+		Path: "/add",
+	}
+
+	sut := DownloadHandler{ro}
+	err := sut.render(appVars, w, r)
+	assert.Nil(err)
+	assert.Equal("422 Unprocessable Entity", w.Result().Status)
+}
```
Comment (hawx) [inline internal/server/download_test.go:75]: could add another test for `api.StatusError`

File: internal/server/download.go [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+
+	err := h.Client().Download(ctx, reportType, reportJournalType, reportScheduleType, reportAccountType, reportDebtType, dateOfTransaction, dateFrom, dateTo, email)
+
+	if err != nil {
+		var (
+			valErr model.ValidationError
+			stErr  api.StatusError
+		)
+		if errors.As(err, &valErr) {
+			data := AppVars{Errors: util.RenameErrors(valErr.Errors)}
```
Comment (hawx) [inline internal/server/download.go:39]: do you want this to ignore the `v AppVars` that is passed in?

File: internal/server/get_downloads.go [hunk 22]
```diff
@@ -1,10 +1,16 @@
 package server
 
 import (
+	"github.com/opg-sirius-finance-admin/internal/model"
 	"net/http"
 )
 
 type GetDownloadsVars struct {
+	ReportsTypes        *[]model.ReportsType
```
Comment (hawx) [inline internal/server/get_downloads.go:9]: a slice is already a pointer, so I would change these to `[]` instead of `*[]`

File: internal/util/util/translate.go [hunk 23]
```diff
@@ -0,0 +1,44 @@
+package util
```
Comment (hawx) [inline internal/util/util/translate.go:1]: `util` isn't a great package name, since it is only used in `server` I'd move this file there and rename the function to `renameErrors`

File: internal/api/api_client_test.go [hunk 24]
```diff
@@ -0,0 +1,53 @@
...
+		Request:    req,
+	}
+
+	err := newStatusError(resp)
+
+	assert.Equal(t, "POST /some/url returned 418", err.Error())
+	assert.Equal(t, err, err.Data())
+}
+
+func SetUpTest() *MockClient {
```
[the remark this answers is not in the document]
Comment (Nicholassully) [reply internal/api/api_client_test.go]: done

File: internal/api/download.go (hunk 6 above)
Comment (Nicholassully) [reply internal/api/download.go]: done

File: internal/api/download.go (hunk 7 above)
Comment (Nicholassully) [reply internal/api/download.go]: done

File: internal/api/api_client.go (hunk 9 above)
Comment (Nicholassully) [reply internal/api/api_client.go]: done

File: internal/api/api_client_test.go (hunk 10 above)
Comment (Nicholassully) [reply internal/api/api_client_test.go:12]: done

File: internal/api/download_test.go (hunk 12 above)
Comment (Nicholassully) [reply internal/api/download_test.go]: done

File: internal/model/date.go (hunk 13 above)
Comment (Nicholassully) [reply internal/model/date.go:9]: Need this in tests not changed the capital letter

File: internal/model/date.go (hunk 14 above)
Comment (Nicholassully) [reply internal/model/date.go:15]: We are sure that this will not get bad data

File: internal/model/date_test.go (hunk 15 above)
Comment (Nicholassully) [reply internal/model/date_test.go]: done

File: internal/model/download.go (hunk 16 above)
Comment (Nicholassully) [reply internal/model/download.go]: done

File: internal/server/download.go (hunk 19 above)
Comment (Nicholassully) [reply internal/server/download.go]: done

File: internal/server/download_test.go (hunk 20 above)
Comment (Nicholassully) [reply internal/server/download_test.go:75]: done

File: internal/server/download.go (hunk 21 above)
Comment (Nicholassully) [reply internal/server/download.go]: Yes we dont want the v AppVars

File: internal/server/get_downloads.go (hunk 22 above)
Comment (Nicholassully) [reply internal/server/get_downloads.go]: done

File: internal/util/util/translate.go (hunk 23 above)
Comment (Nicholassully) [reply internal/util/util/translate.go]: done

File: internal/util/util/translate.go [hunk 25]
```diff
@@ -0,0 +1,44 @@
...
+		"Date":             pair{"Date", "Please select the report date"},
+		"date-in-the-past": pair{"Date", "The report date must be today or in the past"},
+	},
+	"FromDate": {
+		"FromDate": pair{"FromDate", "Date From must be before Date To"},
+	},
+	"ToDate": {
+		"ToDate": pair{"ToDate", "Date To must be after Date From"},
+	},
+}
```
[the remark this answers is not in the document]
Comment (Nicholassully) [reply internal/util/util/translate.go]: Need this as its the same error structure as Sirius

File: internal/model/date_test.go [hunk 26]
```diff
@@ -0,0 +1,175 @@
...
+func TestNewDate(t *testing.T) {
+	want := Date{Time: time.Date(2020, 12, 31, 0, 0, 0, 0, time.UTC)}
+	assert.Equal(t, want, NewDate("31/12/2020"))
+
+	defer func() {
+		if r := recover(); r == nil {
+			t.Errorf("NewDate should panic with incorrect date format")
+		}
+	}()
+	NewDate("12/31/2020") // wrong format should trigger a panic
```
Comment (denialanderror) [inline internal/model/date_test.go:24]: If this is something that can never happen (as per your earlier comment), there shouldn't be a test for it. If it can happen, we shouldn't panic, and should either return an error, or ignore it or return some default.

File: internal/server/server.go [hunk 27]
```diff
@@ -44,3 +52,24 @@ func New(logger *slog.Logger, client ApiClient, templates map[string]*template.T
...
+
+	if r.Method == http.MethodGet {
+		if cookie, err := r.Cookie("XSRF-TOKEN"); err == nil {
+			token, _ = url.QueryUnescape(cookie.Value)
+		}
+	} else {
+		token = r.FormValue("xsrfToken")
+	}
+
+	clientId, _ := strconv.Atoi(r.PathValue("clientId"))
```
Comment (denialanderror) [inline internal/server/server.go:67]: Why is this here? There is no "client" in Finance Admin.

**Review by denialanderror — CHANGES_REQUESTED** [review]
The enums will need json un/marshalling functions, otherwise they will be sent across the wire as integers

File: internal/model/date_test.go (hunk 26 above)
Comment (Nicholassully) [reply internal/model/date_test.go]: done

File: internal/server/server.go (hunk 27 above)
Comment (Nicholassully) [reply internal/server/server.go]: done

**Review by jackgoodby — APPROVED** [review]
